src/mysky/encrypted_files.ts

```diff
@@ -0,0 +1,293 @@
+const ENCRYPTION_NONCE_LENGTH = 24;
+
+/**
+ * The length of the overhead introduced by encryption.
+ */
+const ENCRYPTION_OVERHEAD_LENGTH = 16;
+
+/**
+ * The length of the share-able path seed.
+ */
+export const ENCRYPTION_PATH_SEED_LENGTH = 32;
+
+export type EncryptedJSONResponse = {
+  data: JsonData | null;
+};
+
+type DerivationPathObject = {
+  pathSeed: Uint8Array;
+  directory: boolean;
+  name: string;
+};
+
+type EncryptedFileMetadata = {
+  // 8-bit uint.
+  version: number;
+};
+
+/**
+ * Decrypts the given bytes as an encrypted JSON file.
+ *
+ * @param data - The given raw bytes.
+ * @param key - The encryption key.
+ * @returns - The JSON data and metadata.
+ * @throws - Will throw if the bytes could not be decrypted.
+ */
+export function decryptJSONFile(data: Uint8Array, key: Uint8Array): JsonFullData {
+  validateUint8Array("data", data, "parameter");
+  validateUint8ArrayLen("key", key, "parameter", ENCRYPTION_KEY_LENGTH);
+
+  // Extract the nonce.
+  const nonce = data.slice(0, ENCRYPTION_NONCE_LENGTH);
+
+  // Decrypt the non-nonce part of the data.
+  let decryptedBytes = secretbox.open(data.slice(ENCRYPTION_NONCE_LENGTH), nonce, key);
+  if (!decryptedBytes) {
+    throw new Error("Could not decrypt given encrypted JSON file");
+  }
+
+  // Trim the 0-byte padding off the end.
+  let paddingIndex = decryptedBytes.length;
+  while (paddingIndex > 0 && decryptedBytes[paddingIndex - 1] === 0) {
+    paddingIndex--;
+  }
+  decryptedBytes = decryptedBytes.slice(0, paddingIndex);
+
+  // Extract the metadata.
+  const metadataBytes = decryptedBytes.slice(0, ENCRYPTION_METADATA_LENGTH);
+  decryptedBytes = decryptedBytes.slice(ENCRYPTION_METADATA_LENGTH);
+  const metadata = decodeEncryptedFileMetadata(metadataBytes);
+
+  // Parse the final decrypted message as json.
+  const json = JSON.parse(uint8ArrayToStringUtf8(decryptedBytes));
+  return { _data: json, _v: metadata.version };
+}
+
+/**
+ * Encrypts the given JSON data and metadata.
+ *
+ * @param fullData - The given JSON data and metadata.
+ * @param key - The encryption key.
+ * @returns - The encrypted data.
+ */
+export function encryptJSONFile(fullData: JsonFullData, key: Uint8Array): Uint8Array {
+  const { _data, _v } = fullData;
+  validateObject("fullData._data", _data, "parameter");
+  validateNumber("fullData._v", _v, "parameter");
+  validateUint8ArrayLen("key", key, "parameter", ENCRYPTION_KEY_LENGTH);
+
+  // Stringify the json and convert to bytes.
+  let data = stringToUint8ArrayUtf8(JSON.stringify(_data));
+
+  // Prepend the metadata.
+  const metadata: EncryptedFileMetadata = { version: _v };
+  const metadataBytes = encodeEncryptedFileMetadata(metadata);
+  data = new Uint8Array([...metadataBytes, ...data]);
+
+  // Add padding so that the final size will be a padded block. The overhead will be added by encryption and we add the nonce at the end.
+  const finalSize = padFileSize(data.length) - ENCRYPTION_OVERHEAD_LENGTH - ENCRYPTION_NONCE_LENGTH;
```

Comment:
Noticing this padding happens "before encryption" which makes sense to me, but doesn't seem to follow spec. I don't know how to avoid this, but wanted to mention it in case it's a security thing I don't understand.

> To prevent analysis that can occur by looking at the sizes of files, all encrypted files will be padded to the nearest “pad block”, after encryption.

Reply:
Okay, that should be fixed.